### PR TITLE
Some sample unitPreferences data for discussing output format.

### PR DESCRIPTION
**Pull request by hugovdm**
Let's discuss the desired output format with this limited sample, as we did with unitConversion sample data.

I realised the "geq" value also has to be in the output XPath, not in a value, since it is needed for distinguishing e.g. precision-increment/10 from precision-increment/50 for length/road-travel.

### CODE REVIEW

File: icu4c/source/data/misc/supplementalData.txt [hunk 1]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
+            "road-travel"{
+                001{
+                    kilometer{
+                        geq:0.975{""}
+                    }
+                    meter{
+                        geq:300{
+                            skeleton{"precision-increment/50"}
+                        }
+                        skeleton{"precision-increment/10"}
```
Comment (hugovdm) [inline icu4c/source/data/misc/supplementalData.txt:35801]: I'm not satisfied with this variable depth of skeleton. Maybe we need a "geq:-inf" - then we can also drop the "geq:" prefix. If we do this, then all entries will have a "-inf" (or equivalent) before further formatting- and meta-data.

Maybe I should rather take this to a Google Doc to quickly present alternatives?

File: icu4c/source/data/misc/supplementalData.txt [hunk 2]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
-        }
-        UA{
-            concentr-blood-glucose{"millimole-per-liter"}
-        }
-        UG{
-            concentr-blood-glucose{"millimole-per-liter"}
+        "area"{
+            "geograph"{
+                001{
+                    square-kilometer{""}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: The data types should be the same.  Here, the unit is a string, but below, it is a table.  You should pick one, probably table.  You can declare this as an empty table by appending `:table`:

```
square-kilometer:table{}
```

File: icu4c/source/data/misc/supplementalData.txt [hunk 3]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
-            volume-vehicle-fuel{"gallon"}
+        "duration"{
+            "music-track"{
+                001{
+                    minute+second{""}
+                }
+            }
+            "person-age"{
+                001{
+                    year-person+month-person{""}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Shouldn't this one have a `geq` with bigger values being expressed only in years without months?

File: icu4c/source/data/misc/supplementalData.txt [hunk 4]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
+            "road"{
+                001{
+                    kilometer{""}
+                }
+                GB{
+                    mile{""}
+                }
+                SE{
+                    mile-scandinavian{
+                        alt{"informal"}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Question for @macchiati: does alt="informal" imply that the fallback chain should be used or not?  For example, if SE has foot for small road distance (not alt), and alt="informal" mile-scandanavian, if the alt is requested, do we use feet or not?

File: icu4c/source/data/misc/supplementalData.txt [hunk 5]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
+                    }
+                    US{
+                        foot{""}
+                    }
+                }
+            }
+            "road-travel"{
+                001{
+                    kilometer{
+                        geq:0.975{""}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: The 0.975 should be the string value, like this:

```
geq{"0.975"}
```

File: icu4c/source/data/misc/supplementalData.txt [hunk 6]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
+            }
+            "tv-program"{
+                001{
+                    minute+second{""}
+                }
+            }
+            _default{
+                001{
+                    year{
+                        geq:1{""}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Here and elsewhere: the value should be inside the quotes, like

```
geq{"1"}
```

File: icu4c/source/data/misc/supplementalData.txt [hunk 7]
```diff
@@ -35597,398 +35597,297 @@ supplementalData:table(nofallback){
...
+                    }
+                    month{
+                        geq:1{""}
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Possibly leave out the `geq` if it is 1?

File: icu4c/source/data/misc/supplementalData.txt (hunk 3 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: I guess this becomes a design question - so "foot+inch" would always output both, ala "0 feet 3 inches", and thus we will use geq to switch to "inch" only. Done.

File: icu4c/source/data/misc/supplementalData.txt (hunk 2 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: From a discussion with hagbard it sounds like paths have to be predictable (can't be iterated over unless we use an :array), so unit identifiers need to move back into values.

I've reworked this sample, what do you think of the enumerated approach?

File: icu4c/source/data/misc/supplementalData.txt (hunk 6 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: I need something in the path to disambiguate. Turns out geq values can't be in the path at all (nevermind misusing the colon in my first draft) unless other values inform the code what values to look for. Done.

File: icu4c/source/data/misc/supplementalData.txt (hunk 7 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: Done.

File: icu4c/source/data/misc/supplementalData.txt (hunk 5 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: Ah... currently not a string yet. There's an :int type but I don't find something like :float or :double - so floats have to be passed as strings?

**Review by hugovdm — COMMENTED** [review]
Fixing mix of tabs spaces.

**Comment by hugovdm** [comment]
ldml2icu also has a <FIFO> feature, which can produce what we have here minus the integer enumerations (so the same, just `s/\s*\d{/\s*{/` (i.e. dropping the single digits, the reset stays the same). I'll try to produce that.

File: icu4c/source/data/misc/supplementalData.txt [hunk 8]
```diff
@@ -35597,398 +35597,385 @@ supplementalData:table(nofallback){
...
+                }
+                US{
+                    0{
+                        unit{"mile"}
+                    }
+                }
+                // DEMO: Current hack for _small, proposing we drop this info and simply use geq:
+                _small{
+                    001{
+                        0{
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Can you make this an array instead of a 0/1/2/... table?

Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: How does this look to you? This was actually generated by <FIFO> rules. And there are additional changes pulled in which I didn't manually remove again - makes it easier to regenerate. (Output is generated on a cldr branch that's based on sffc/units-staging.)

File: icu4c/source/data/misc/supplementalData.txt [hunk 9]
```diff
@@ -7860,6 +7860,7 @@ supplementalData:table(nofallback){
                 "Qaaj~p",
             }
             special{
+                "Aran",
```
Comment (sffc) [inline icu4c/source/data/misc/supplementalData.txt]: Let's keep this branch clean; don't check in changes unrelated to units

**Review by sffc — COMMENTED** [review]
One small comment, but other than that, looks great!

**Comment by macchiati** [comment]
In the doc https://docs.google.com/document/d/1JxId6OP3oh4UEwHdhZ0qOjkTExMwk72xtKBLBlohDtQ/edit, we have the structure:

```
<unitPreferences category="length" usage="road">
  <unitPreference regions="CA IN US" unit="foot" />
  <unitPreference regions="CA IN US" geq=”67” unit="yard" /> <!-- 200 feet -->
  <unitPreference regions="CA IN US" geq="0.568" unit="mile" /> <!-- 1000 yards -->
...
```

That doesn't seem to correspond to what you have, but would be more like:
```
length {
  road {
    CA {
      0 {foot} // 0 = default; we would test the absolute value so that works
      61.2648 {yard} // measures converted to baseUnit for length, ie meters
      914.10739 {mile}
     }
   ...
  person {
   ...
area {
...
```

**Comment by sffc** [comment]
The structure Hugo has here I think is appropriate for the implementation.  The converter spec will map from the XML structure to the "JSON" structure.  For example, we need to use vecs rather than tables because the maps will be sorted in undesirable ways.

**Comment by macchiati** [comment]
That makes sense for the bottom level, changing this into an array of pairs.
{
      0 {foot} // 0 = default; we would test the absolute value so that works
      61.2648 {yard} // measures converted to baseUnit for length, ie meters
      914.10739 {mile}
}

My main question is on the higher levels; it feels like the unitQuantity is the highest level, and then usage then region. That also matches more closely what we have in CLDR, making it easier to reformat (though that is a minor consideration).

=====

Second issue is that for some other data we end up sharing items, to cut down on space. Something like the following. Now, since this is just data sharing and the ICU format is changeable, we could do this or other things later if it makes sense.

length {
  road {
    CA { set1 }
    US { set1}


sets {
  set1 {
      0 {foot} // 0 = default; we would test the absolute value so that works
      61.2648 {yard} // measures converted to baseUnit for length, ie meters
      914.10739 {mile}
  }
  set2 {
....

**Comment by sffc** [comment]
I'm open to the idea of sets to help organize this a bit better.  I feel like it might be premature optimization, though.  There isn't really very much data here; it is all in supplemental, so we don't get the additional locale multiplier for size either.

**Comment by macchiati** [comment]
Makes sense.

File: icu4c/source/data/misc/supplementalData.txt (hunk 9 above)
Comment (hugovdm) [reply icu4c/source/data/misc/supplementalData.txt]: Done.

**Comment by hugovdm** [comment]
I've removed the unrelated changes. This is therefore good sample data in the format we seem to have reached consensus on. Next up, I'll try to add back all the data that I've dropped (due to being in the old format).

I think this branch can be merged (git merge) by Younies, if he wants to start using this sample data. If in the future we change the format again, I'd be happy to help fix all the unit tests that break as a consequence.

**Review by sffc — APPROVED** [review]

**Comment by sffc** [comment]
Let's get this into units-staging so that we can move forward.

**Pull request merged**